test/SILOptimizer/destroy_barrier_unit.sil

```diff
@@ -0,0 +1,172 @@
+// RUN: %target-sil-opt -test-runner %s -enable-experimental-feature MoveOnlyEnumDeinits -o /dev/null | %FileCheck %s
+
+// REQUIRES: swift_feature_MoveOnlyEnumDeinits
+
+sil_stage canonical
+
+import Builtin
+import Swift
+
+struct S1: ~Copyable {
+  deinit
+}
+
+struct S2: ~Copyable {
+  deinit
+}
+
+struct S3<T: ~Copyable>: ~Copyable {
+  @_hasStorage let t: T
+  deinit
+}
+
+enum E<T: ~Copyable>: ~Copyable {
+  case A(Int)
+  case B
+  case C(T)
+  deinit
+}
+
+struct StrWithoutDeinit: ~Copyable {
+  @_hasStorage var a: S1
+  @_hasStorage var b: S2
+  @_hasStorage let c: Int
+}
+
+struct S4: ~Copyable {
+  @_hasStorage var a: S1
+  @_hasStorage var b: S2
+}
+
+struct S5: ~Copyable {
+  @_hasStorage let a: Int
+  @_hasStorage let b: AnyObject
+}
+
+struct UnknownDeinit: ~Copyable {
+  deinit
+}
+
+sil @s1_deinit : $@convention(method) (@owned S1) -> () {
+[global: read]
+}
+
+sil @s2_deinit : $@convention(method) (@owned S2) -> () {
+[global: write]
+}
+
+sil @s3_deinit : $@convention(method) <T> (@in S3<T>) -> () {
+[global: read]
+}
+
+sil @e_deinit : $@convention(method) (@owned S4) -> () {
+[global: read]
+}
+
+sil_moveonlydeinit S1 {
+  @s1_deinit
+}
+
+sil_moveonlydeinit S2 {
+  @s2_deinit
+}
+
+sil_moveonlydeinit S3 {
+  @s3_deinit
+}
+
+sil_moveonlydeinit E {
+  @e_deinit
+}
+
+sil_global @g : $Int
+
+sil [ossa] @global_load_store : $@convention(thin) (@inout_aliasable Int, @guaranteed S1, @guaranteed S2, @guaranteed S3<S1>, @guaranteed S3<S2>, @guaranteed S4, @guaranteed S5, @guaranteed E<S4>, @guaranteed E<Int>, @guaranteed UnknownDeinit) -> () {
```

Comment:
It's challenging to read these semantics because the types only appear in the argument list. The mismatch between argument number and type number makes it especially hard (you need three editor windows). If each type appeared next to its `CHECK-LABEL` either as a move_value or in the CHECK comment, then it would be easy. I'm thinking of people who refer back to this to understand the rules.